[Backtest] Including intraday with trend

msperetto opened the pull request:
This PR will add all functionality to include the capability of running intraday backtests with the possibility of optimize the trend classes ass well

msperetto committed:
Add manager to handle intraday backtests including trend;

Similar to intraday manager, but including onTrend() checks when opening long and short positions;

msperetto committed:
Add json config file for intraday trend

msperetto committed:
Add all dependencies and code to run new backtest for intraday w/ trend

ltoledozt commented on main/backtest/app.py:
@msperetto outro ponto de melhoria:
- Salvarmos na base se trata-se de operações Long ou Short.

ltoledozt commented on main/backtest/app.py:
```diff
@@ -52,6 +53,7 @@ def __init__(self):
             Json_type.INTRADAY: "main/backtest/resources/intraday_params.json",
             Json_type.STRATEGY: "main/backtest/resources/strategy_params.json",
             Json_type.PORTFOLIO: "main/backtest/resources/portfolio_params.json"
```
@msperetto Reparei que esse arquivo está com vários erros de compilação! 
Por exemplo, ta faltando uma `,` no final dessa linha.

Estranho seu editor python não falar isso!! Eu uso o Visual Code e ele sempre dá um highlight em erros de tipo, ou mesmo quando falta alguma referencia!

ltoledozt commented on main/backtest/app.py:
```diff
@@ -182,6 +187,34 @@ def run_intraday_optimization(self, bt):
+            self.trade_buy_classes,
+            self.filter_sell_classes,
+            self.trigger_sell_classes,
+            self.trade_sell_classes
```
Faltando uma `,` no final dessa linha!

msperetto replied:
Meu VScode tá meio estranho mesmo... reiniciei o PC agora e voltou a aparecer os highlights...

luciozt committed:
Refactor encapsulating logic to run backtest for a specific perior

It should be functionally equivalent.
It will be useful in subsequent commits.

luciozt committed:
Add ability to run backtest in different date ranges

Introduced BacktestSplitMode enum with options FULL, MONTHLY, and CUSTOM_DAYS to allow flexible partitioning of backtest periods.

Updated logic to generate sub-ranges accordingly.
Added support for configuration via 'split_mode' and optional 'split_days' parameter.

msperetto committed:
temp change configuration to run specific backtest

msperetto committed:
Refactor extended start Time calculation

As the longest trend indicator can impact backtest for short periods, the start date should start N days before the real start date, where N is the amount of time of the longest trend indicator;

msperetto committed:
Fix START and END dates to be datetime, instead of strings;

msperetto committed:
Add INTRADAY_TREND to backtest manager selector;

ltoledozt commented on main/backtest/app.py:
```diff
@@ -131,21 +137,36 @@ def get_optimization_params(self):
             "trend_interval": self.trend_interval
         }
 
+    def _get_longest_trend_indicator_from_params(self):
```
@msperetto Acho q esse metodo foi incluido no commit errado! 
Pelo git, diz que foi no commit:  `temp change configuration to run specific backtest`
O correto seria o: `Refactor extended start Time calculation`.

Confere?

msperetto replied:
É.. tá no commit errado mesmo, quando inclui o trecho acima do código veio esse junto e nem percebi...

Tem um jeito de fazer esse ajuste facilmente?

ltoledozt commented on main/common/management.py:
```diff
@@ -45,9 +49,18 @@ def dict_to_params(dict):
-    diff_size = time_intervals_to_seconds(strategy.trend_interval)*(strategy.get_biggest_trend_interval()+TREND_INTERVAL_CALC_MARGIN)
-    start_date = datetime.now() - timedelta(seconds=diff_size)
-    return start_date.strftime('%d.%m.%Y')
+def calc_start_date(strategy = None, trend_period = None, longest_indicator = None, base_start_date = datetime.now()):
```
@msperetto Alguma chance dessa alteração impactar produção?

msperetto replied:
Eu fiz o código pra não impactar.

Em produção, nós passamos sempre uma strategy, que não é o caso desse momento do backtest, onde passamos só a trend.
O IF abaixo trata esses 2 cenários. Sempre que utilizarmos esse método, passamos ou strategy, ou a trend e ele funciona pros 2 casos

ltoledozt commented on main/backtest/app.py:
```diff
@@ -284,7 +307,54 @@ def get_backtest_manager(self):
+        # Iterate through each date range and run the backtest.
+        # This will allow us to run the backtest for each month or custom days as defined in the config.
+        for range_start, range_end in date_ranges:
+            extended_start_time = management.calc_start_date(trend_period=self.trend_interval, longest_indicator=self._get_longest_trend_indicator_from_params(), base_start_date=range_start)
```
Uma dica: evitar chamar funções dentro de argumento de funções.

Nesse caso, seria interessante definir uma variável que chama `self._get_longest_trend_indicator_from_params()`. E ai passa o resultado na `management.calc_start_date`.

```
longest_indicator = self._get_longest_trend_indicator_from_params()
extended_start_time = management.calc_start_date(trend_period=self.trend_interval, longest_indicator=longest_indicator, base_start_date=range_start)
```

msperetto replied:
Legal!

msperetto committed:
Fix logic to reset triggers when buy or sell confirmation are True

msperetto committed:
Refactor passing method param as variable instead of fuction call

msperetto committed:
Add side (long or short) information to backtest report

ltoledozt commented on main/backtest/app.py:
```diff
@@ -92,10 +99,10 @@ def create_strategy_dict(self, strategy_keys):
     # Runs the logic to save a row in the Optimization_test table.
     # It will save only if the global "should_save_report" flag is True.
-    def save_report(self, stats, strategy_class = ""):
+    def save_report(self, stats, strategy_class = "", side = None):
```
Ao invés de passar `side` como parâmetro e ter que passar essa informação como argumento em todos lugares que chamam, ficaria mais simples definir isso como uma variável dentro do próprio `save_report` method.

```
def save_report(self, stats, strategy_class = ""):
    side = self.config["operation_type"].name
    if self.config["should_save_report"]:
        cut_long_string = str(stats["_strategy"]).find(",filter_buy_class")
        db.insert_report(self.pair, str(self.interval), stats, str(stats["_strategy"])[:cut_long_string]+")", self.period_label, self.trend_interval, strategy_class, side=side)
```

msperetto replied:
Tinha feito daquele outro modo pro caso de rodarmos várias estratégias de long e short simultaneamente (a ideia do PORTIFOLIO), mas incluindo um "else None" esse jeito fica melhor

ltoledozt replied:
Interessante! Bom ponto! 👍
